### DISCUSSION TITLE
'Model creation failed (Model not available in objective at level 1)'

### DISCUSSION THREAD

**Discussion by valuestay** [yalmip/YALMIP#1379] [Problems using YALMIP] [answered]
Dear Professor,
It's the first time that I use yalmip, when I try to optimize the following objective, it appears the following error. I want to know what this error means, how should I modify the code?
code:
```
g_d1=complex(randn(1,4), randn(1,4));
g_0=complex(randn(1,20), randn(1,20));
g_1=complex(randn(20,4), randn(20,4));
a_t=randn(4,1);

theta0_r = sdpvar(20,1,'full','complex');

Constraints = [abs(theta0_r) <= 1];
Objective=(abs(g_d1+g_0*diag(theta0_r)*g_1).^2)*a_t;
sol = optimize(Constraints,Objective);     

if sol.problem == 0
    solution = value(Objective);
else
    disp('出错了!');
    sol.info
    yalmiperror(sol.problem)
end
```
errror:
```
出错了!

ans =

    'Model creation failed (Model not available in objective at level 1)'


ans =

    'Model creation failed '
```

**Comment by johanlofberg** [comment]
abs on complex expressions means norm, and then using that inside a quadratic will violate convexity propagation rules. Hence you will have to write the quadratic explicitly, something like

```
e = g_d1+g_0*diag(theta0_r)*g_1
objective = e'*(a_t.*e)
```
or model things using auxilliary variables and epigraphs, something like

```
e = sdpvar(4,1)
Model = abs(g_d1+g_0*diag(theta0_r)*g_1) <= e
objective = sum(e.*a_t.*e)

```

**Reply by valuestay** [reply]
Thanks for your reply.In fact, the above problem is a subproblem, and the full optimization objective is shown below. After I modified the code according to your guidance, a time-out problem sometimes occurred when I ran the following code,and I'm not sure whether the results are optimal. Is there any way to shorten the run time, or set maximum time of iterations (I don't know which solver I'm using)?
code：
[parameter.zip](https://github.com/yalmip/YALMIP/files/14507421/parameter.zip)

```
unzip('parameter.zip');
load('parameter.mat');
theta0_r = sdpvar(M,1,'full','complex');
theta0_t = sdpvar(M,1,'full','complex');
t0_r = sdpvar(1,1);
t0_t = sdpvar(1,1);
tk_r = sdpvar(k,1);
tk_t = sdpvar(k,1);
e1=g_d1+conj(g_0)*diag(theta0_r)*g_1;
e2=g_d2+conj(g_0)*diag(theta0_t)*g_2;

Constraints = [abs(theta0_r) <= 1,abs(theta0_t) <= 1,...    
    t0_r + t0_t + sum(tk_r) + sum(tk_t) == T,...   
    tk_t' * log(1 + t0_r*P*a_t.*(diag(conj(e1))*e1') ./ tk_t)>=R0,...
    tk_r' * log(1 + t0_t*P*a_r.*(diag(conj(e2))*e2') ./ tk_r)>=R0];

Objective = -tk_t' * log(1 + t0_r*P*a_t.*(diag(conj(e1))*e1') ./ tk_t) - ...
                    tk_r' * log(1 + t0_t*P*a_r.*(diag(conj(e2))*e2') ./ tk_r);
                
sol = optimize(Constraints,Objective);         

if sol.problem == 0
    solution = -value(Objective);
else
    disp('出错了!');
    sol.info
    yalmiperror(sol.problem)
end
```
error：
```
······  
  782    2991   -1.012613e+00    1.690e-02    2.345e+03    6.748e-03
  783    2993   -1.012613e+00    1.690e-02    2.345e+03    1.860e-02
  784    2999   -1.012611e+00    1.690e-02    2.345e+03    5.369e-03

Solver stopped prematurely.

fmincon stopped because it exceeded the function evaluation limit,
options.MaxFunctionEvaluations = 3.000000e+03.

出错了!

ans =

    'Maximum iterations or time limit exceeded (FMINCON)'


ans =

    'Maximum iterations or time limit exceeded '
```
By the way,The original constraint was `abs(theta0_r) == 1`, but an error occurred:` 'Model creation failed (Model not available in constraint #1 at level 1)'`. So I changed it to `abs(theta0_r) <= 1`. did that make an influence to the final result?

**Reply by johanlofberg** [reply, marked as answer]
You are setting up a large nonconvex nonlinear program, so you are simply at the mercy of the performance of fmincon

Model looks bad too, you have things like stuff/t >= constant which should be written as stuff >= t*constant to avoid singularity-inducing nonlinearity

You also might want to impose sparsity to make computations inside fmincon, and yalmip faster by not doing e = stuff and then create e'*e but instead define new variable e with constraint e==stuff, thus creating very sparse quadratics (if stuff is a dense operator of the decision variables)

Yes changing f(x)==0 to f(x)<=0 will obviously create a different problem with possible different result if it is optimal to not satisfy the constraint. You cannot use norm (which abs is) in equalities as norm is convexity preserving (and requires socp modelling which is impossible here as the model is way beyond socp-representable) and a nonlinear equality never is convex. you have to derive the quadratic counterpart
